### PR TITLE
Add horizon finder into BBH executable

### PR DESCRIPTION
**Pull request by markscheel**
## Proposed changes

Adds horizon finder into BBH executable.

I don't think CI actually runs this executable, so it should be tested before merging. (thats why I made it a draft)

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

**Comment by wthrowe** [comment]
I've managed to write a control system measurement based on this that at least compiles.  (I don't have the data to run it either.)  I think it should be enough for at least the systems that only use the centers.

**Comment by markscheel** [comment]
Thanks @wthrowe ! I don't have the data to run this either, so I was hoping that @geoffrey4444 would check out my branch and try it.

File: src/Evolution/Executables/GeneralizedHarmonic/EvolveGhBinaryBlackHole.hpp [hunk 1]
```diff
@@ -176,6 +179,92 @@ struct EvolutionMetavars {
...
+    // Find horizon in grid frame.
+    using vars_to_interpolate_to_target = tmpl::list<
+        gr::Tags::SpatialMetric<volume_dim, ::Frame::Grid, DataVector>,
+        gr::Tags::InverseSpatialMetric<volume_dim, ::Frame::Grid>,
+        gr::Tags::ExtrinsicCurvature<volume_dim, ::Frame::Grid>,
+        gr::Tags::SpatialChristoffelSecondKind<volume_dim, ::Frame::Grid>>;
+    using compute_vars_to_interpolate = ah::ComputeHorizonVolumeQuantities;
+    // Observe horizon in grid frame too.
+    using tags_to_observe = tmpl::list<
+        StrahlkorperGr::Tags::AreaCompute<::Frame::Grid>,
```
Comment (geoffrey4444) [inline src/Evolution/Executables/GeneralizedHarmonic/EvolveGhBinaryBlackHole.hpp]: <optional> Since you typically want to do the same thing on AhA and AhB, would be nice to define some aliases once and then use them for both AhA and AhB, wherever aliases in AhA and AhB structs are exactly the same

File: tests/InputFiles/GeneralizedHarmonic/BinaryBlackHole.yaml [hunk 2]
```diff
@@ -194,6 +204,30 @@ Observers:
...
 
+ApparentHorizons:
+  AhA:
+    InitialGuess:
+      Lmax: 10
+      Radius: 2.2
+      Center: [*XCoordA, 0.0, 0.0]
+    FastFlow: &DefaultFastFlow
+      Flow: Fast
+      Alpha: 1.0
```
Comment (geoffrey4444) [inline tests/InputFiles/GeneralizedHarmonic/BinaryBlackHole.yaml:216]: Shouldn't you set these options for AhB as well?

File: tests/InputFiles/GeneralizedHarmonic/BinaryBlackHole.yaml [hunk 3]
```diff
@@ -194,6 +204,30 @@ Observers:
...
+      Beta: 0.5
+      AbsTol: 1e-12
+      TruncationTol: 1e-2
+      DivergenceTol: 1.2
+      DivergenceIter: 5
+      MaxIts: 100
+    Verbosity: Verbose
+  AhB:
```
Comment (geoffrey4444) [inline tests/InputFiles/GeneralizedHarmonic/BinaryBlackHole.yaml:224]: Why are Alpha, Beta, etc. not options to set for AhB?

**Review by geoffrey4444 — CHANGES_REQUESTED** [review]
I have some questions about the code. Also, when I tried running it on a BBH (`ocean:/home/geoffrey/aps2021/inspiral_size_attempt/bbh_exec_attempt/ah/fail1`), I got an error from a root finder complaining about bracketing:

~~~
Line: 61 of /home/geoffrey/Codes/spectre/spectre/src/NumericalAlgorithms/RootFinding/RootBracketing.hpp
Function: std::array<double, 4> RootFinder::bracketing_detail::bracket_by_contracting(const std::vector<double> &, const std::vector<std::optional<double> > &, const Functor &, const size_t) [Functor = (lambda at /home/geoffrey/Codes/spectre/spectre/src/NumericalAlgorithms/RootFinding/RootBracketing.hpp:334:9)]
Too many iterations in bracket_by_contracting. Either refine the initial range/guess or increase max_level.
############ ERROR ############
~~~

Maybe it's trying to invert a map at a place where it's not invertible?

**Comment by geoffrey4444** [comment]
Yes, CI only tests that the input file options work, not that the executable runs successfully

File: tests/InputFiles/GeneralizedHarmonic/BinaryBlackHole.yaml (hunk 2 above)
Comment (markscheel) [reply tests/InputFiles/GeneralizedHarmonic/BinaryBlackHole.yaml:216]: Yes, they are set.  Thats what the `&DefaultFastFlow` and the later `*DefaultFastFlow` does.

File: tests/InputFiles/GeneralizedHarmonic/BinaryBlackHole.yaml (hunk 3 above)
Comment (markscheel) [reply tests/InputFiles/GeneralizedHarmonic/BinaryBlackHole.yaml:224]: They are.  That's what `*DefaultFastFlow` does.  Look up "Yaml aliases" and "Yaml anchors".

**Comment by markscheel** [comment]
Fixup commit adds aliases for things common to AhA and AhB.

Executable should now work provided #3594 is also used; the bracketing error that @geoffrey4444 mentions above is fixed by #3594 .

**Comment by geoffrey4444** [comment]
I will test this again and let you know if it works for me

**Review by geoffrey4444 — CHANGES_REQUESTED** [review]
I rebased this on latest develop, and I verified that it the AH finder now works successfully!

I'm happy to approve, but first please rebase on latest develop, and verify that the test input file works? In the current version, the domain is out of date. I put the input file I used successfully here:

~~~
wheeler:/home/geoffrey/temp/BinaryBlackHole.yaml
ocean:/home/geoffrey/aps2021/inspiral_bbh_exec_attempt/spectre_run/v2ah/Run/BinaryBlackHole.yaml
~~~

Edit: please squash

**Comment by markscheel** [comment]
Squashed and rebased, and added another commit that updates the input file.

@geoffrey4444 In your example input file there were lines in EventsAndTriggers that were commented out; when I changed the input file I just removed those.

**Review by geoffrey4444 — DISMISSED** [review]

**Comment by markscheel** [comment]
The failing test is Unit.PointwiseFunctions.AnalyticSolutions.Elasticity.HalfSpaceMirror which was not touched in this PR.

File: src/Evolution/Executables/GeneralizedHarmonic/EvolveGhBinaryBlackHole.hpp [hunk 4]
```diff
@@ -177,6 +180,73 @@ struct EvolutionMetavars {
...
+                     volume_dim, ::Frame::Grid, DataVector>,
+                 StrahlkorperTags::NormalOneFormCompute<::Frame::Grid>,
+                 StrahlkorperTags::UnitNormalOneFormCompute<::Frame::Grid>,
+                 StrahlkorperTags::UnitNormalVectorCompute<::Frame::Grid>,
+                 StrahlkorperTags::GradUnitNormalOneFormCompute<::Frame::Grid>,
+                 StrahlkorperTags::ExtrinsicCurvatureCompute<::Frame::Grid>,
+                 StrahlkorperGr::Tags::SpinFunctionCompute<::Frame::Grid>>,
+      horizons_tags_to_observe>;
+
+  struct AhA {
```
Comment (nilsvu) [inline src/Evolution/Executables/GeneralizedHarmonic/EvolveGhBinaryBlackHole.hpp:210]: (optional) AFAICT the `AhA` and `AhB` classes are identical. How about this:

```cpp
template <typename Label>
struct ApparentHorizon { ... };
struct AhA {};
struct AhB {};
using interpolation_target_tags = tmpl::list<ApparentHorizon<AhA>, ApparentHorizon<AhB>>;
```

File: tests/InputFiles/GeneralizedHarmonic/BinaryBlackHole.yaml [hunk 5]
```diff
@@ -190,6 +190,16 @@ EventsAndTriggers:
...
+  ? Slabs:
+      EvenlySpaced:
+        Interval: 1
+        Offset: 0
+  : - AhA
+  ? Slabs:
+      EvenlySpaced:
+        Interval: 1
+        Offset: 0
+  : - AhB
```
Comment (nilsvu) [inline tests/InputFiles/GeneralizedHarmonic/BinaryBlackHole.yaml]:
```suggestion
  ? Slabs:
       EvenlySpaced:
         Interval: 1
         Offset: 0
   : - AhA
     - AhB
```

File: tests/InputFiles/GeneralizedHarmonic/BinaryBlackHole.yaml [hunk 6]
```diff
@@ -136,27 +136,9 @@ Filtering:
   ExpFilter0:
     Alpha: 36.0
     HalfPower: 210
-    DisableForDebugging: false
+    DisableForDebugging: true
```
Comment (nilsvu) [inline tests/InputFiles/GeneralizedHarmonic/BinaryBlackHole.yaml:139]: This says "for debugging". Did you intend to change this?

File: tests/InputFiles/GeneralizedHarmonic/BinaryBlackHole.yaml [hunk 7]
```diff
@@ -136,27 +136,9 @@ Filtering:
...
-      - TimeCompares:
-          Comparison: GreaterThan
-          Value: 0.1
-  : - ChangeSlabSize:
-        DelayChange: 0
-        StepChoosers:
-          - Cfl:
-              SafetyFactor: 0.8
-          - Increase:
-              Factor: 1.2
```
Comment (nilsvu) [inline tests/InputFiles/GeneralizedHarmonic/BinaryBlackHole.yaml]: Just for my information, why did you remove the change in slab size?

**Review by nilsvu — CHANGES_REQUESTED** [review]
Please squash directly if you end up making changes.

File: tests/InputFiles/GeneralizedHarmonic/BinaryBlackHole.yaml (hunk 7 above)
Comment (markscheel) [reply tests/InputFiles/GeneralizedHarmonic/BinaryBlackHole.yaml]: I removed it because @geoffrey4444 did;  I just copied his input files that he said that worked.

File: tests/InputFiles/GeneralizedHarmonic/BinaryBlackHole.yaml (hunk 6 above)
Comment (markscheel) [reply tests/InputFiles/GeneralizedHarmonic/BinaryBlackHole.yaml:139]: I could change it;  @geoffrey4444 wanted me to use his input file, and this was present in his input file.

**Comment by nilsvu** [comment]
you can just squash this @markscheel :D

File: src/Evolution/Executables/GeneralizedHarmonic/EvolveGhBinaryBlackHole.hpp (hunk 4 above)
Comment (markscheel) [reply src/Evolution/Executables/GeneralizedHarmonic/EvolveGhBinaryBlackHole.hpp:210]: If I do this, then the code that determines the input file options would need to change too.  Currently the `AhA` in the input file is the same as the InterpolationTargetTag, so if the InterpolationTargetTag changes to `ApparentHorizon<AhA>` then you would need to say something like `ApparentHorizon<AhA>` in the input file (if that would work), or change the code for getting the input file tag from the InterpolationTargetTag.

It seems simpler to me if the name in the input file is the same as the name of the tag in the metavariables.

**Comment by markscheel** [comment]
Made an input file fix and squashed.  Some other input file things are a result of me copying a working input file from @geoffrey4444 , so perhaps the slab size and the debugging flag can be restored?

File: src/Evolution/Executables/GeneralizedHarmonic/EvolveGhBinaryBlackHole.hpp (hunk 4 above)
Comment (nilsvu) [reply src/Evolution/Executables/GeneralizedHarmonic/EvolveGhBinaryBlackHole.hpp:210]: I think the code that determines the input file name should use `Options::name<>()`, then it's just a matter of defining a static `name()` function. I can't find the place in the code where the input file name is determined though to check what it does. Either way, I'm fine leaving this as-is.

**Review by nilsvu — DISMISSED** [review]

File: src/Evolution/Executables/GeneralizedHarmonic/EvolveGhBinaryBlackHole.hpp (hunk 4 above)
Comment (wthrowe) [reply src/Evolution/Executables/GeneralizedHarmonic/EvolveGhBinaryBlackHole.hpp:210]: I did basically that in #3620: https://github.com/sxs-collaboration/spectre/pull/3620/files#diff-48693954adbf700f151d18b3160a43b05928950c7e46e12579da1142708f6ea4R54 (`Interpolator` is an interpolation target tag.)

Comment (nilsvu) [reply src/Evolution/Executables/GeneralizedHarmonic/EvolveGhBinaryBlackHole.hpp:210]: @wthrowe are you requesting a change here?

**Comment by nilsvu** [comment]
@geoffrey4444 could you comment briefly on the input file comments above?

File: tests/InputFiles/GeneralizedHarmonic/BinaryBlackHole.yaml (hunk 6 above)
Comment (geoffrey4444) [reply tests/InputFiles/GeneralizedHarmonic/BinaryBlackHole.yaml:139]: I no longer recommend filtering for BBH runs, since it doesn't seem necessary in my tests, but I'm unsure whether to remove it altogether, so I prefer disabling in yaml for now

File: tests/InputFiles/GeneralizedHarmonic/BinaryBlackHole.yaml (hunk 7 above)
Comment (geoffrey4444) [reply tests/InputFiles/GeneralizedHarmonic/BinaryBlackHole.yaml]: You can revert and squash this change...I was doing fixed time step runs, but I think this adaptive slab chooser is fine for the test input file

**Comment by geoffrey4444** [comment]
> @geoffrey4444 could you comment briefly on the input file comments above?

I did above :) please let me know if you need more input from me!

**Review by geoffrey4444 — DISMISSED** [review]

**Comment by markscheel** [comment]
I put back the ChangeSlabSize stuff in the input file and squashed.

**Review by geoffrey4444 — APPROVED** [review]

**Review by nilsvu — APPROVED** [review]
